### ISSUE TITLE
TypeError: Cannot read property 'circle' of null

### ISSUE THREAD

**Issue by micooz** [d3/d3-voronoi#24]
I encountered an issue that special input data can crash **voronoi.polygons**.

**Reproduce Demo**

```js
var d3 = require("d3")

var width = 400;
var height = 400;

// change width and height to 100 works well
// var width = 100;
// var height = 100;

var x = d3.scaleLinear();
var y = d3.scaleLinear();

x.domain([0, 1]).range([0, width]).nice();
y.domain([0, 1]).range([height, 0]).nice();

var voronoi = d3.voronoi()
  .extent([[-1, -1], [width + 1, height + 1]])
  .x((pt) => x(pt.x))
  .y((pt) => y(pt.y));

// TypeError: Cannot read property 'circle' of null
voronoi.polygons([
  {x: 0.9999994485078688, y: 0},
  {x: 0.9999994512180275, y: 0}
]);
```

I thought it was my scale range problem, but the following code works:

```
voronoi.polygons([
  {x: 0.1, y: 0},
  {x: 0.2, y: 0}
]);
```

https://runkit.com/micooz/595cabe5640c440013382792

**Comment by mbostock** [comment]
Here’s a reduction:

```js
var voronoi = d3.voronoi();

voronoi([
  [399.999779, 0],
  [399.999780, 0]
]);
```

In short the problem is [the rounding applied during construction](https://github.com/d3/d3-voronoi/blob/v1.1.2/src/voronoi.js#L12) isn’t enough to prevent certain numerical instabilities; these two points are approximately, but not exactly, coincident. This is probably not something I can fix since this code is ported from Raymond Hill’s implementation of Steven Fortune’s algorithm. However, it might be fixable if this library is rewritten to use Vladimir Agafonkin’s [delaunator](https://github.com/mapbox/delaunator), so that it first computes the Delaunay triangulation and then derives the dual Voronoi diagram. I believe this is generally considered the more numerically stable approach, and it’ll probably be faster…

**Comment by Fil** [comment]
https://beta.observablehq.com/d/2e32a5a801b9ed02 shows this is probably fixed now in d3-delaunay

I'm saying “probably” because d3-delaunay can't yet handle cases with n=2 points, cf https://github.com/d3/d3-delaunay/issues/19

**Comment by garrickf** [comment]
Does d3-delaunay allow for a mapping from the Voronoi polygons to data points (with the other attributes in our original data), like d3-voronoi does? I wanted to use this because of the ability to make it easier to hover over points in a scatter plot, but I'm also running into the issue of some of my data points being approximately coincident.

**Comment by mbostock** [comment]
> Does d3-delaunay allow for a mapping from the Voronoi polygons to data points

Yes. Cell *i* in the Voronoi diagram corresponds to input point *i*.

**Comment by garrickf** [comment]
> > Does d3-delaunay allow for a mapping from the Voronoi polygons to data points
>
> Yes. Cell _i_ in the Voronoi diagram corresponds to input point _i_.

Thanks for the quick reply! I think my issue might be different: the number of cells I get is less than the number of points I pass in, so perhaps some of my points are _actually_ coincident? I found a way to create a mapping for at least some of my points that have cells, which will probably suffice for hovering over information. Thanks for your help!

**Comment by Fil** [comment]
It will indeed be fixed by https://github.com/d3/d3-delaunay/pull/64
